mlflow/tracing/export/inference_table.py

```diff
@@ -0,0 +1,94 @@
+    """
+    Get the completed trace data from the buffer. This method is used in
+    the Databricks model serving so please be careful when modifying it.
+    """
+    return _TRACE_BUFFER.get(request_id)
+
+
+# For Inference Table, we use special TTLCache to store the finished traces
+# so that they can be retrieved by the Databricks model serving. This needs
+# to be a different from the queue-based buffer used by MLflow fluent API as
+# the access is not index-based but key (request ID) based.
+_TRACE_BUFFER = TTLCache(
+    maxsize=MLFLOW_TRACE_BUFFER_TTL_SECONDS.get(),
+    ttl=MLFLOW_TRACE_BUFFER_MAX_SIZE.get(),
+)
+
+
+class InferenceTableSpanExporter(SpanExporter):
+    """
+    An exporter implementation that logs the traces to Inference Table.
+
+    Currently the Inference Table does not use collector to receive the traces,
+    but rather actively fetches the trace during the prediction process. In the
+    future, we may consider using collector-based approach and this exporter should
+    send the traces instead of storing them in the buffer.
+    """
+
+    def __init__(self):
+        self._trace_manager = InMemoryTraceManager.get_instance()
+
+    def export(self, root_spans: Sequence[ReadableSpan]):
+        """
+        Export the spans to MLflow backend.
+
+        Args:
+            spans: A sequence of OpenTelemetry ReadableSpan objects to be exported.
+                Only root spans for each trace are passed to this method.
+        """
+        for span in root_spans:
+            if span._parent is not None:
+                _logger.debug("Received a non-root span. Skipping export.")
+                continue
+
+            trace = self._trace_manager.pop_trace(span.context.trace_id)
+            if trace is None:
+                _logger.debug(f"Trace for span {span} not found. Skipping export.")
+                continue
+
+            # Add the trace to the in-memory buffer so it can be retrieved by upstream
+            _TRACE_BUFFER[trace.info.request_id] = {
+                TRACE_SCHEMA_VERSION_KEY: 2,
+                "start_timestamp": self.nanoseconds_to_datetime(span._start_time),
+                "end_timestamp": self.nanoseconds_to_datetime(span._end_time),
+                "spans": [self.format_spans(span) for span in trace.data.spans],
+            }
+
+    def format_spans(self, mlflow_span: LiveSpan) -> Dict[str, Any]:
+        """
+        Format the MLflow span to the format that can be stored in the Inference Table.
+
+        The schema is mostly the same, but the attributes field is stored as a JSON
+        string instead of a dictionary. Therefore, the attributes are converted from
+        Dict[str, str(json)] to str(json).
+        """
+        span_dict = mlflow_span.to_dict()
+        attributes = span_dict["attributes"]
+        # deserialize each attribute value and then serialize the whole dictionary
+        attributes = {k: json.loads(v) for k, v in attributes.items()}
```

Comment:
Should we iterate over this dict and handle JSONDecodeError issues by retaining a string encoding over the reference?
The attributes can be specified by the user, and I'm curious if a value that is supplied with malformed JSON contained within it could present an issue here.

Reply:
Generally we don't have to worry about that, as we are the only one doing that validation when they set attributes in spans ([code](https://github.com/mlflow/mlflow/blob/tracing/mlflow/entities/span.py#L446-L454)). One edge case would be when user directly access the underlying OTel span somehow and put attribute, which I think is fine to raise exception. WDYT?

Reply:
That's pretty much what I was getting at. Since we're directly interfacing with those public APIs internally when creating spans, there's nothing stopping a user from just using those APIs as well (or if they're porting over some existing implementation that uses OTeL to MLflow, having a safe handling of those inputs for custom trace telemetry might save some headaches)

Reply:
Ah that makes sense! Yea keep the original value in case of a decoding error makes sense.